accepted/2020/binary-data/binary-data.md

```diff
@@ -0,0 +1,195 @@
+}
+```
+
+## Requirements
+
+### Goals
+
+* Make it easy for people to construct bytes from common data types (byte
+  arrays, spans, memories, string, JSON serialized objects)
+* Make it easy for people to convert bytes to common the same representations
+* Provide opinionated defaults in order to reduce complexity
+    - String encoding is assumed to be UTF8
+    - Object serialization is done with `System.Text.Json`, with specific options
+
+### Non-Goals
+
+* Allocation free zero-copying semantics. The primary goal is convenience and
+  usability, not perf. Performance critical scenarios should be enabled by
+  providing overloads with the existing low-level concepts, such as span.
+
+## Design
+
+We're concerned with hard coding the type to `System.Text.Json`, but we
+acknowledge that defaults are very valuable. If we need to support other
+serializers we can provide additional arguments or APIs.
+
+This type can't live in corlib because it would depend on `System.Text.Json`
+(and whatever future serializers it needs to support). Hence, most areas in the
+BCL wouldn't be able to take `BinaryData`, but technologies above the BCL could
+(Microsoft.Extensions, ASP.NET Core, Azure SDK etc). While that might be
+unfortunate, we believe it's acceptable because most of the usability gains
+would be in higher levels anyway.
+
+Thus, we shouldn't think of this as an exchange type in the BCL. We already have
+exchange types; the point of this type is unify them by providing conversions;
+the point isn't to provide yet another type. Thus, we don't see the need to, for
+example, provide new virtual methods on Stream.
+
+We could decide to model the constructor that performs JSON serialization as a
+factory method. Combined with a hypothetical feature *extension statics* we
+could make the type more low level than serialization. An alternative design is
+to replace constructors and factories with instance members (e.g. `SetBytes`
+`SetObjectFromJson`) which would throw when the underlying array is non-null.
+That prevents mutation and uses a simple pattern
+
+```C#
+var data = new BinaryData();
+data.SetBytes(...);
+```
+
+This would also allow us to use regular extension methods to split out JSON
+serialization.
+
+We need to balance layering concerns with usability. The primary goal of this
+type is to provide usability via conveniences. If we don't make the type super
+usable, the feature has failed. While the design with extension methods feels
+quite appealing we need to verify that it realizes these goal. If we can't, we
+should go back to the design of using constructors and factory methods, even if
+that reduces our flexibility.
+
+### API design with extension methods
+
+```C#
+namespace System
+{
+    public readonly struct BinaryData
+    {
+        public void SetBytes(ReadOnlyMemory<byte> data);
+        public void SetBytes(ReadOnlySpan<byte> data);
+        public void SetBytes(byte[] data);
+        public void SetStream(Stream stream);
```

Comment:
who owns the stream, or does this essentially loop over the stream and read all the bytes and store it internally?

Reply:
All of the setters are data sinks (or the factory methods/ctors in the other variant).  The Streams aren't held as references, they're drained into a buffer.

This ends up being important, so that things like ToBytes don't need Async variants.

Reply:
Should we be concerned about the potential for misuse of this particular method? E.g. passing NetworkStream's whose length doesn't fit in a single buffer.

Reply:
> passing NetworkStream's whose length doesn't fit in a single buffer.

This API is akin to the various `ReadToEnd` methods we have throughout the framework. I don't think those methods have historically been misused.

Reply:
> The Streams aren't held as references, they're drained into a buffer.

This seemed the most likely, but it would be nice to spell this out in the XML documentation for the method.